File: files/300-399/327/CH-327.py
import pandas as pd
import re

path = "300-399/327/CH-327 Column Splitting.xlsx"
input = pd.read_excel(path, usecols="B", nrows = 8, skiprows = 1)
test = pd.read_excel(path, usecols= "D:E", nrows = 8, skiprows = 1)

level_pattern = "|".join(map(re.escape, (levels := ["Upper Ground", "Ground", "Under Ground"])))
zone_pattern = "|".join(map(re.escape, (levels :=  ["West", "East", "North", "South", "South East", "North West"])))

df = input.copy()
df["Level"] = df["Info"].apply(lambda info: re.findall(level_pattern, str(info)))
df["Zone"] = df["Info"].apply(
    lambda info: (
        " ".join(zones) if len((zones := re.findall(zone_pattern, str(info)))) > 1
        else zones[0] if zones
        else None
    )
)
df = df.explode("Level")
df = df.drop(columns=["Info"])

print(df.equals(test)) # True

File: files/300-399/327/CH-327.R
library(tidyverse)
library(readxl)

path <- "300-399/327/CH-327 Column Splitting.xlsx"
input <- read_excel(path, range = "B2:B9")
test  <- read_excel(path, range = "D2:E9")

levels = c("Upper Ground", "Ground", "Under Ground")
Zones = c("West", "East", "North", "South", "South East", "North West")

result = input %>%
  rowwise() %>%
  mutate(
    Level = str_extract_all(Info, paste0(levels, collapse = "|")),
    Zone = str_extract_all(Info, paste0(Zones, collapse = "|")),
    Zone = if (length(Zone) > 1) paste(Zone, collapse = " ") else as.character(Zone)
  ) %>%
  ungroup() %>%
  unnest(Level) %>%
  select(-Info)

all.equal(result, test)
# [1] TRUE
